src/ray/core_worker/store_provider/memory_store/memory_store.h

```diff
@@ -0,0 +1,94 @@
+#ifndef RAY_CORE_WORKER_MEMORY_STORE_H
+#define RAY_CORE_WORKER_MEMORY_STORE_H
+
+#include "ray/common/buffer.h"
+#include "ray/common/id.h"
+#include "ray/common/status.h"
+#include "ray/core_worker/common.h"
+
+namespace ray {
+
+class ReferencedRayObject;
+class GetOrWaitRequest;
+class CoreWorkerMemoryStore;
+
+/// A class that represents an object in memory store.
+class ObjectEntry {
+ public:
+  ObjectEntry(const ObjectID &object_id, const RayObject &object);
+
+  std::shared_ptr<RayObject> GetObject() const { return object_; }
+  int IncreaseRefcnt() { return ++refcnt_; }
+  int DecreaseRefcnt() { return --refcnt_; }
+  int Refcnt() { return refcnt_; }
+
+  std::shared_ptr<ReferencedRayObject> CreateReferencedObject(
+      std::shared_ptr<CoreWorkerMemoryStore> provider);
+
+ private:
+  const ObjectID object_id_;
+  std::shared_ptr<RayObject> object_;
+  std::atomic<uint16_t> refcnt_;
+};
+
+/// This cache currently contains unconsumed objects, which are the objects
+/// that have been `put` but haven't been `get`.
+class EvictionCache {
```

Comment:
For the local memory store, we shouldn't use LRU eviction and use reference counting instead. It is easy to implement in a single process (in Python using reference counting, in Java I'm not sure, what's the right way to do it there?) and will make sure nothing gets deleted while it is needed.

Reply:
Here we are already using reference counting, plus a limit on the total size of the objects.

here are how it works currently:
1. Use reference counting on the objects returned from `ray.get`, when the reference count drops to zero, the object is automatically deleted. This means objects will get deleted immediately after they are used.
2. The store has a limit of the size of total objects, and it maintains a list of objects that haven't been used, these objects can be evicted in FIFO order when total object size reaches the limit. 

As I understand your suggestion is to remove rule 2 above, so that objects remain in store as long as it haven't been `ray.get` from worker right?  A possible issue is that user calls a large number of remote calls but doesn't call `ray.get`, then the store memory can grow unlimited. Should we handle this case?

Reply:
In my suggestion, 1) would delete the object when there is no `ObjectID` referencing it any more. In that case it is safe to delete. Deleting when the gotten object goes out of scope is not safe, because there might still be an object id hanging around and the user might call ray.get on it later.

Concerning 2., if we are in that case, we cannot guarantee correctness, since the object might still be in use. I think this is better implemented by supporting actor memory limits like in https://github.com/ray-project/ray/pull/5226 and giving an error if the actor uses more memory.

Reply:
Thanks for the explanations.

For suggestion 1) Yes I agree this would be more desirable. I think this is a generic problem which is not particular to in-memory store, if an object is evicted from plasma store then the subsequent `ray.get` would not find it from store and would then try to reconstruct the object, if it cannot reconstruct then it fails with `object not reconstructable" error. I think for now we can handle this case with a "object evicted" error if it's not found in store, later we can implement the suggested logic in python/java to make it safe. I'll add a TODO in the code.

For 2) Yes that sounds good to me. I'll remove the eviction policy then so unconsumed objects will not be removed by default.

Reply:
Please remove the whole eviction policy and also implement the RayObject proposal from below. It will make the PR way simpler.

Reply:
Yep, thanks for the suggestions, appreciate that. 

For the eviction policy, yes, will remove.

For RayObject, will simplify C++ worker side accordingly, and file a later PR to address it on python side after python worker is migrated.